backends/cuda-reg/ceed-cuda-reg.c

```diff
@@ -36,8 +36,7 @@ static int CeedInit_Cuda_reg(const char *resource, Ceed ceed) {
 
   int currentDeviceID;
   ierr = cudaGetDevice(&currentDeviceID); CeedChk_Cu(ceed,ierr);
-  if (currentDeviceID!=deviceID)
-  {
+  if (currentDeviceID!=deviceID) {
     ierr = cudaSetDevice(deviceID); CeedChk_Cu(ceed,ierr);
```

Comment:
I recently understood how the `CudaContext` works. The problem is that the runtime API uses only the Primary context, whereas the driver API can use other contexts (not only the Primary). This `if` statement was made to avoid destroying the context used by another application. However, I should check that this `if` statement is really needed. Nonetheless, I think that this is unfortunately insufficient in the general case, since the application using libCEED might be using the driver API, in which case calling libCEED would switch the context... (to use the Primary one) which would result in the GPU not accepting the GPU pointers from the application using libCEED (coming from a context different from the current Primary context).
I think that for now it is fine to assume that we can use the Primary context, but we should discuss about using the driver API to avoid this kind of issues in the future. We could require that appications using libCEED to use the Primary context (even with the driver API).

Reply:
Thanks. On browsing the [docs](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1gb02d4c850eb16f861fe5a29682cc90ba), I think we want `cuCtxPushCurrent()` on entry to public libCEED interfaces (upon reaching CUDA backends) and `cuCtxPopCurrent()` on exit so that libCEED activities on the device do not conflict with the user or with other libraries (all of which may have async operations overlapping on the same or different devices).

Reply:
The problem with that it that it requires to use the driver API everywhere, since using the runtime API anywhere automatically switches the context to the Primary context, thus the change goes far beyond this one line.
PS: This is why I was calling `cudaFree(0)` at the beginning of the function `compile`, this forces the creation/switching to the Primary context.

Reply:
Is the problem that the driver API is too cumbersome for general use? How do "official" libraries with asynchronous interfaces do it?

Reply:
The driver API is more challenging to use, I'm not comfortable with it, but this can be a goal for the future. The use of the runtime API is relatively limited in the cuda backends, mostly for things like `cudaMemcpy`. Since we use nvrtc we already use the driver API for a big part of the cuda backends. MFEM is using the runtime API for instance, MFEM was using the some of the driver API initially but change for the total runtime API for the 4.0 release, Occa is using the driver API I believe, Umpire is using the runtime API (and the developers highly recommend to do the same)... I don't see that as a high priority, but this is definitely something worth keeping in mind.

Reply:
Thanks. By "official", I was thinking of NVIDIA-developed open source packages. Technical decisions can change a lot when NVIDIA needs to pay people to support it.

For example, this is pinned on their GitHub and seems to mostly use the driver API. https://github.com/NVIDIA/DALI

Reply:
As far as I know, switching contexts is an expensive operation and therefore should be avoided.